### PR TITLE
Prs/ace graphs

### PR DESCRIPTION
**Pull request by rattus128**
https://github.com/Comfy-Org/ComfyUI/pull/15694/changes (minor rework)

Minor rework of ^^ to minimize the conditional branching logic in llama implementation. Use the generic scatter approach consistently including in gemma and unify on just the original FixedKV class with functionality appropriately extended for the seqlen over batch variance.

Example test conditions:

Linux, RTX5090, CPU underclocked to 2GHz
Ace-step 1.5 turbo XL template (qwen 1.7b ckpt)

image

Before:

```
LM sampling: 100%|██████████| 600/600 [00:13<00:00, 43.96it/s]
```

After:

```
LM sampling: 100%|██████████| 600/600 [00:03<00:00, 156.92it/s]
```

Outputs checked ✅ 

Regression tests:

Gemma 4 ✅ 
Minimax music ✅

### CODE REVIEW

File: comfy/text_encoders/ace15.py [hunk 1]
```diff
@@ -4,9 +4,29 @@
...
 
+def _audio_logits(model, x, audio_start, audio_end, eos_token=None):
+    input = x[:, -1:]
+    module = model.embed_tokens
+
+    offload_stream = None
+    if module.comfy_cast_weights:
+        weight, _, offload_stream = comfy.ops.cast_bias_weight(module, input, offloadable=True)
+    else:
+        weight = module.weight.to(x)
```
[the remark this answers is not in the document]
Comment (rattus128) [reply comfy/text_encoders/ace15.py:19]: You need to prove that the llama contract is fully in play. Please check the actual ace15 checkpoints on huggingface and if you cannot find a checkpoint that reproduces the issue, please withdraw.

File: comfy/text_encoders/llama.py [hunk 2]
```diff
@@ -824,6 +832,10 @@ def forward(self, x, attention_mask=None, embeds=None, num_tokens=None, intermed
         past_len = 0
         if past_key_values is not None and len(past_key_values) > 0:
             past_len = self.get_past_len(past_key_values)
+        fixed_kv = past_key_values is not None and len(past_key_values) > 0 and isinstance(past_key_values[0], FixedKV)
+        fixed_kv_decode = fixed_kv and past_len > 0 and seq_len == 1
+        if fixed_kv_decode:
+            attention_mask = None
```
[the remark this answers is not in the document]
Comment (rattus128) [reply comfy/text_encoders/llama.py:838]: agenting counter opinion from review of review:

No—the [CodeRabbit major comment](https://github.com/Comfy-Org/ComfyUI/pull/15698#discussion_r3801814212) is incorrect for ACE15’s left-padding path.

Example: padded width 10, valid length 6.

- Prefill valid keys receive RoPE positions 4…9.
- They are compacted into cache slots 0…5, but retain those RoPE rotations.
- Decode must therefore use RoPE position 10 (past_len), while writing to cache slot 6 (seqlen).
- Using seqlen == 6 for RoPE would mismatch the cached keys.

FixedKV.prepare() correctly keeps these concepts separate:

- position = seqlen: physical cache write slot
- seqlen += 1: attended cache extent
- index/past_len: logical padded RoPE timeline

CodeRabbit’s fix would only be valid if prefill also assigned compact per-row position IDs. This path does not. I’d mark the comment as a false positive.

**Review by comfyanonymous — APPROVED** [review]
